File: ios/HolyDay/Services/Support/TipService.swift
//
//  TipService.swift
//  HolyDay
//
//  Created by Matthias Cadet on 14/05/2026.
//

import Foundation
import Observation
import RevenueCat

@Observable
final class TipService {
  static let shared = TipService()

  // Derived from RevenueCat CustomerInfo; falls back to cached UserDefaults value
  private(set) var hasTipped: Bool = UserDefaults.standard.bool(forKey: "holyday.hasTipped") {
    didSet { UserDefaults.standard.set(hasTipped, forKey: "holyday.hasTipped") }
  }

  // Stored as index+1 so 0 means "never purchased" (UserDefaults returns 0 for missing keys)
  private var tipTierIndexStored: Int = UserDefaults.standard.integer(
    forKey: "holyday.highestTipIndex")
  {
    didSet { UserDefaults.standard.set(tipTierIndexStored, forKey: "holyday.highestTipIndex") }
  }

  private(set) var tipsOffering: Offering?

  // Palier dérivé du rang de prix dans l'offering (le moins cher = palier le plus bas).
  // Robuste aux changements d'identifiants ou de prix des produits sur l'App Store, contrairement
  // à une correspondance basée sur le nom du produit (« tip_small/medium/large »).
  private var tierByProductId: [String: SupporterTier] = [:]

  var supporterTier: SupporterTier? {
    guard hasTipped else { return nil }
    return SupporterTier(rawValue: tipTierIndexStored - 1)
  }

  private init() {
    Task { await refreshCustomerInfo() }
  }

  func refreshCustomerInfo() async {
    async let customerInfoTask: CustomerInfo? = try? Purchases.shared.customerInfo()
    async let offeringsTask: Offerings? = try? Purchases.shared.offerings()

    // L'offering doit être appliqué AVANT les transactions : la correspondance produit → palier
    // est dérivée du rang de prix de l'offering, donc la carte doit déjà être construite.
    let offs = await offeringsTask
    tipsOffering = offs?.offering(identifier: RevenueCatConfig.offeringId)
    rebuildTierMap()

    if let info = await customerInfoTask {
      applyCustomerInfo(info)
    }
  }

  // Carte produit → palier construite par rang de prix croissant de l'offering.
  private func rebuildTierMap() {
    let ranked =
      (tipsOffering?.availablePackages ?? [])
      .sorted { $0.storeProduct.price < $1.storeProduct.price }
    var map: [String: SupporterTier] = [:]
    for (index, package) in ranked.enumerated() {
      guard let tier = SupporterTier(rawValue: index) else { break }
      map[package.storeProduct.productIdentifier] = tier
    }
    tierByProductId = map
  }

  // Repli sur la correspondance par nom tant que l'offering n'est pas chargé.
  func tier(for productIdentifier: String) -> SupporterTier? {
    tierByProductId[productIdentifier] ?? SupporterTier.tier(for: productIdentifier)
  }

  // Enregistre directement le palier acheté. Les dons sont des produits CONSOMMABLES et, en mode
  // utilisateur anonyme, n'apparaissent plus dans CustomerInfo après l'achat (ils ne sont pas
  // conservés sur le reçu du store — cf. doc RevenueCat). Le badge est donc persisté localement
  // depuis le palier connu au moment de l'achat, sans dépendre d'un aller-retour CustomerInfo.
  func recordPurchase(tier: SupporterTier) {
    hasTipped = true
    tipTierIndexStored = tier.rawValue + 1
  }

  #if DEBUG
    // Bascule le badge supporter pour tester les écrans gated des deux côtés, sans passage au store.
    func debugSetSupporter(_ enabled: Bool) {
      hasTipped = enabled
      tipTierIndexStored = enabled ? 1 : 0
    }
  #endif

  func applyCustomerInfo(_ info: CustomerInfo) {
    let entitlementActive = info.entitlements[RevenueCatConfig.entitlementId]?.isActive == true
    let hasTransactions = !info.nonSubscriptions.isEmpty

    // Purement ADDITIF : CustomerInfo ne sert qu'à (ré)confirmer le badge, jamais à l'effacer.
    // Pour des consommables anonymes, un CustomerInfo vide est l'état normal (et non un
    // remboursement) — l'effacer figerait le badge à « jamais visible » dès le 1er refresh/lancement.
    guard entitlementActive || hasTransactions else { return }
    hasTipped = true
    if hasTransactions {
      updateTier(from: info.nonSubscriptions)
    } else if tipTierIndexStored == 0 {
      // Droit actif sans transaction connue (ex. restauration) : au moins le palier minimal.
      tipTierIndexStored = 1
    }
  }

  // Le badge reflète le don le PLUS RÉCENT (et non le palier le plus élevé jamais atteint) : on
  // réaffecte sans `max` pour qu'il puisse aussi redescendre si le dernier don est d'un palier
  // inférieur. Sinon, un don élevé ponctuel figerait le badge sur ce palier indéfiniment.
  private func updateTier(from transactions: [NonSubscriptionTransaction]) {
    let latestTier =
      transactions
      .max { $0.purchaseDate < $1.purchaseDate }
      .flatMap { tier(for: $0.productIdentifier) }
    // Repli : un achat est actif mais le palier reste introuvable → palier minimal.
    tipTierIndexStored = (latestTier?.rawValue ?? 0) + 1
  }
}

File: android/app/src/main/java/com/matthiascadet/holyday/service/TipService.kt
package com.matthiascadet.holyday.service

import com.matthiascadet.holyday.data.model.SupporterTier
import com.matthiascadet.holyday.data.prefs.AppPreferences
import com.matthiascadet.holyday.ui.theme.RevenueCatConfig
import com.revenuecat.purchases.CustomerInfo
import com.revenuecat.purchases.Offering
import com.revenuecat.purchases.Purchases
import com.revenuecat.purchases.awaitCustomerInfo
import com.revenuecat.purchases.awaitOfferings
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

/**
 * Équivalent de `TipService` iOS : dons consommables via RevenueCat, badge dérivé du palier
 * le plus récent (jamais du plus élevé jamais atteint — voir `applyCustomerInfo`).
 */
object TipService {
    private const val HAS_TIPPED_KEY = "holyday.hasTipped"
    private const val HIGHEST_TIP_INDEX_KEY = "holyday.highestTipIndex"

    private val _hasTipped = MutableStateFlow(false)
    val hasTipped: StateFlow<Boolean> = _hasTipped.asStateFlow()

    private val _tipsOffering = MutableStateFlow<Offering?>(null)
    val tipsOffering: StateFlow<Offering?> = _tipsOffering.asStateFlow()

    // Stocké en index+1 : 0 = "jamais acheté" (SharedPreferences renvoie 0 pour une clé absente).
    private var tipTierIndexStored: Int = 0

    private var tierByProductId: Map<String, SupporterTier> = emptyMap()

    val supporterTier: SupporterTier?
        get() {
            if (!_hasTipped.value) return null
            return SupporterTier.entries.find { it.rank == tipTierIndexStored - 1 }
        }

    fun init() {
        val prefs = AppPreferences.raw
        _hasTipped.value = prefs.getBoolean(HAS_TIPPED_KEY, false)
        tipTierIndexStored = prefs.getInt(HIGHEST_TIP_INDEX_KEY, 0)
        CoroutineScope(Dispatchers.IO).launch { refreshCustomerInfo() }
    }

    suspend fun refreshCustomerInfo() {
        val offerings = runCatching { Purchases.sharedInstance.awaitOfferings() }.getOrNull()
        _tipsOffering.value = offerings?.getOffering(RevenueCatConfig.OFFERING_ID)
        rebuildTierMap()

        val info = runCatching { Purchases.sharedInstance.awaitCustomerInfo() }.getOrNull()
        if (info != null) applyCustomerInfo(info)
    }

    private fun rebuildTierMap() {
        val ranked = (_tipsOffering.value?.availablePackages ?: emptyList())
            .sortedBy { it.product.price.amountMicros }
        val map = mutableMapOf<String, SupporterTier>()
        for ((index, pkg) in ranked.withIndex()) {
            val tier = SupporterTier.entries.find { it.rank == index } ?: break
            map[pkg.product.id] = tier
        }
        tierByProductId = map
    }

    fun tier(for_: String): SupporterTier? = tierByProductId[for_] ?: SupporterTier.forProductIdentifier(for_)

    /**
     * Enregistre directement le palier acheté. Les dons sont des produits consommables et,
     * en mode utilisateur anonyme, n'apparaissent plus dans CustomerInfo après l'achat — le
     * badge est donc persisté localement depuis le palier connu au moment de l'achat.
     */
    fun recordPurchase(tier: SupporterTier) {
        setHasTipped(true)
        setTipTierIndexStored(tier.rank + 1)
    }

    fun debugSetSupporter(enabled: Boolean) {
        setHasTipped(enabled)
        setTipTierIndexStored(if (enabled) 1 else 0)
    }

    /**
     * Purement ADDITIF : CustomerInfo ne sert qu'à (ré)confirmer le badge, jamais à l'effacer.
     * Pour des consommables anonymes, un CustomerInfo vide est l'état normal (pas un
     * remboursement) — l'effacer figerait le badge à "jamais visible" dès le 1er lancement.
     */
    fun applyCustomerInfo(info: CustomerInfo) {
        val entitlementActive = info.entitlements[RevenueCatConfig.ENTITLEMENT_ID]?.isActive == true
        val transactions = info.nonSubscriptionTransactions
        val hasTransactions = transactions.isNotEmpty()

        if (!entitlementActive && !hasTransactions) return
        setHasTipped(true)
        if (hasTransactions) {
            updateTier(transactions)
        } else if (tipTierIndexStored == 0) {
            setTipTierIndexStored(1)
        }
    }

    /**
     * Le badge reflète le don le PLUS RÉCENT (pas le palier le plus élevé jamais atteint) : un
     * don élevé ponctuel ne doit pas figer le badge sur ce palier indéfiniment.
     */
    private fun updateTier(transactions: List<com.revenuecat.purchases.models.Transaction>) {
        val latestTier = transactions.maxByOrNull { it.purchaseDate.time }
            ?.let { tier(it.productIdentifier) }
        setTipTierIndexStored((latestTier?.rank ?: 0) + 1)
    }

    private fun setHasTipped(value: Boolean) {
        _hasTipped.value = value
        AppPreferences.raw.edit().putBoolean(HAS_TIPPED_KEY, value).apply()
    }

    private fun setTipTierIndexStored(value: Int) {
        tipTierIndexStored = value
        AppPreferences.raw.edit().putInt(HIGHEST_TIP_INDEX_KEY, value).apply()
    }
}
